Invalid dead code elimination with loops without iterations

<!--

Thanks for opening an issue! To help the Numba team handle your information
efficiently, please first ensure that there is no other issue present that
already describes the issue you have
(search at https://github.com/numba/numba/issues?&q=is%3Aissue).

-->

## Reporting a bug

<!--

Before submitting a bug report please ensure that you can check off these boxes:

-->

- [x] I have tried using the latest released version of Numba (most recent is
  visible in the release notes
  (https://numba.readthedocs.io/en/stable/release-notes-overview.html).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

```python
from numba import njit

@njit()
def jitted(low):
    val = None

    for _ in range(low, 2):
        val = 'garbage'

    if val == None:
        print('jit: loop had no iterations')
    else:
        print('jit: loop had an iteration')

def interpreted(low):
    val = None

    for _ in range(low, 2):
        val = 'garbage'

    if val == None:
        print('interpreter: loop had no iterations')
    else:
        print('interpreter: loop had an iteration')

jitted(5)
interpreted(5)
```

<!--

Please include details of the bug here, including, if applicable, what you
expected to happen!

-->
The jitted version prints that the loop had an iteration, while the interpreted that it did not.
The JIT compiler incorrectly assumes that `val` will be changed and eliminates the code responsible for when `val` is `None`. 

Interestingly enough, this does not happen when instead of `None`/`'garbage'` you do it with `True`/`False`.

> My version of the patch to make it work for this case:
> ```
> % git diff
> diff --git a/numba/cpython/builtins.py b/numba/cpython/builtins.py
> index 1ebb2d778..9aa582478 100644
> --- a/numba/cpython/builtins.py
> +++ b/numba/cpython/builtins.py
> @@ -137,7 +137,7 @@ def gen_non_eq(val):
>              return impl
>          elif a_none ^ b_none:
>              def impl(a, b):
> -                return not val
> +                return a is b
>              return impl
>      return none_equality
> ```
> I didn't check the full testsuite

This fails with:

```
======================================================================
FAIL: test_eq_ne (numba.tests.test_builtins.TestOperatorMixedTypes.test_eq_ne)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/esc/git/numba/numba/tests/test_builtins.py", line 1178, in test_eq_ne
    self.assertPreciseEqual(func.py_func(x, y), func(x, y))
    ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/esc/git/numba/numba/tests/support.py", line 484, in assertPreciseEqual
    self.fail("when comparing %s and %s: %s" % (first, second, failure_msg))
    ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: when comparing True and False: True != False

----------------------------------------------------------------------
Ran 10722 tests in 837.565s

FAILED (failures=1, skipped=629, expected failures=23)
make: *** [test] Error 1
make test  3803.10s user 1198.92s system 596% cpu 13:59.15 total
```

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -1,5 +1,5 @@
 Fix comparison between ``Optional`` and ``None``
--------------------------------------------------
+------------------------------------------------
 
 Comparisons ``==`` (``operator.eq``) and ``!=`` (``operator.ne``)  between
 ``Optional`` and ``None`` are now handled correctly.

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -1,5 +1,5 @@
-Fix comparison between ``Optional(None)`` and ``None``
-------------------------------------------------------
+Fix comparison between ``Optional`` and ``None``
+-------------------------------------------------
 
-Equality comparisons such as ``==`` and ``!=``  between ``Optional(None)`` and
-``None`` are now handled correctly.
+Comparisons ``==`` (``operator.eq``) and ``!=`` (``operator.ne``)  between
+``Optional`` and ``None`` are now handled correctly.

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -2,4 +2,5 @@ Fix comparison between ``Optional`` and ``None``
 ------------------------------------------------
 
 Comparisons ``==`` (``operator.eq``) and ``!=`` (``operator.ne``)  between
-``Optional`` and ``None`` are now handled correctly.
+``Optional`` and ``None`` are now handled correctly. Also, comparisons of any
+type other than ``Optional`` with ``None`` now yield a boolean literal.

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -1,5 +1,5 @@
-Fix and test for 10414
-----------------------
+Fix comparison between ``Optional(None)`` and ``None``
+------------------------------------------------------
 
-Equality comparisons such as `==` and `!=`  between ``Optional(None)`` and
+Equality comparisons such as ``==`` and ``!=``  between ``Optional(None)`` and
 ``None`` are now handled correctly.

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -1,0 +1,5 @@
+Fix and test for 10414
+----------------------
+
+Equality omparisons such as `==` and `!=`  between ``Optional(None)`` and
+``None`` are now handled correctly.

--- a/docs/upcoming_changes/10418.bug_fix.rst
+++ b/docs/upcoming_changes/10418.bug_fix.rst
@@ -1,5 +1,5 @@
 Fix and test for 10414
 ----------------------
 
-Equality omparisons such as `==` and `!=`  between ``Optional(None)`` and
+Equality comparisons such as `==` and `!=`  between ``Optional(None)`` and
 ``None`` are now handled correctly.

--- a/numba/cpython/builtins.py
+++ b/numba/cpython/builtins.py
@@ -137,7 +137,10 @@ def gen_non_eq(val):
             return impl
         elif a_none ^ b_none:
             def impl(a, b):
-                return not val
+                if val:
+                    return a is b
+                else:
+                    return not a is b
             return impl
     return none_equality
 

--- a/numba/cpython/builtins.py
+++ b/numba/cpython/builtins.py
@@ -133,14 +133,8 @@ def gen_non_eq(val):
         b_none = isinstance(b, types.NoneType)
         # Case: both are None.
         if a_none and b_none:
-            if val:
-                # Case None == None --> Always True.
-                def impl(a, b):
-                    return True
-            else:
-                # Case None != None --> Always False.
-                def impl(a, b):
-                    return False
+            def impl(a, b):
+                    return val
             return impl
         # Case: only one is None.
         elif a_none ^ b_none:
@@ -161,15 +155,9 @@ def gen_non_eq(val):
                 return impl
             # Otherwise one is None, the other isn't.
             else:
-                if val:
-                    # Case Any == None --> Always False
-                    def impl(a, b):
-                        return False
-                else:
-                    # Case Any != None --> Always True
-                    def impl(a, b):
-                        return True
-
+                nval = not val
+                def impl(a, b):
+                    return nval
                 return impl
     return none_equality
 

--- a/numba/cpython/builtins.py
+++ b/numba/cpython/builtins.py
@@ -134,7 +134,7 @@ def gen_non_eq(val):
         # Case: both are None.
         if a_none and b_none:
             def impl(a, b):
-                    return val
+                return val
             return impl
         # Case: only one is None.
         elif a_none ^ b_none:

--- a/numba/cpython/builtins.py
+++ b/numba/cpython/builtins.py
@@ -131,17 +131,46 @@ def gen_non_eq(val):
     def none_equality(a, b):
         a_none = isinstance(a, types.NoneType)
         b_none = isinstance(b, types.NoneType)
+        # Case: both are None.
         if a_none and b_none:
-            def impl(a, b):
-                return val
+            if val:
+                # Case None == None --> Always True.
+                def impl(a, b):
+                    return True
+            else:
+                # Case None != None --> Always False.
+                def impl(a, b):
+                    return False
             return impl
+        # Case: only one is None.
         elif a_none ^ b_none:
-            def impl(a, b):
+            a_optional = isinstance(a, types.Optional)
+            b_optional = isinstance(b, types.Optional)
+            # Special case, one optional, needs a runtime check.
+            # Using the implementation for `is` as the distinction between `is`
+            # and `==` is semantically irrelevant in the Numba context. Numba
+            # types can not override `__eq__` like Python objects can so the
+            # two comparisons are semantically equivalent.
+            if a_optional or b_optional:
                 if val:
-                    return a is b
+                    def impl(a, b):
+                        return a is b
                 else:
-                    return not a is b
-            return impl
+                    def impl(a, b):
+                        return not a is b
+                return impl
+            # Otherwise one is None, the other isn't.
+            else:
+                if val:
+                    # Case Any == None --> Always False
+                    def impl(a, b):
+                        return False
+                else:
+                    # Case Any != None --> Always True
+                    def impl(a, b):
+                        return True
+
+                return impl
     return none_equality
 
 overload(operator.eq)(gen_non_eq(True))

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1177,6 +1177,28 @@ class TestOperatorMixedTypes(TestCase):
             for x, y in itertools.product(things, things):
                 self.assertPreciseEqual(func.py_func(x, y), func(x, y))
 
+    def test_10414(self):
+        # test for https://github.com/numba/numba/issues/10414
+        @njit
+        def func_eq():
+            a = None
+            for _ in range(0):
+                a = 'b'
+            # a is now an OptionalType(unicode_type) -- we test that equals
+            # works correctly.
+            return a == None
+        self.assertEqual(func_eq(), func_eq.py_func())
+
+        @njit
+        def func_ne():
+            a = None
+            for _ in range(0):
+                a = 'b'
+            # a is now an OptionalType(unicode_type) -- we test that not equals
+            # works correctly.
+            return a != None
+        self.assertEqual(func_ne(), func_ne.py_func())
+
     def test_cmp(self):
         for opstr in ('gt', 'lt', 'ge', 'le', 'eq', 'ne'):
             op = getattr(operator, opstr)

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1182,8 +1182,10 @@ class TestOperatorMixedTypes(TestCase):
 
         def check(operator):
             for (x, y) in ((None, None), (None, 1), (1, None)):
+
                 def bar(result):
                     pass
+
                 @overload(bar, prefer_literal=True)
                 def ol_bar(result):
                     # compile expected result
@@ -1195,9 +1197,11 @@ class TestOperatorMixedTypes(TestCase):
                     # Assert that result has the correct literal_value
                     self.assertEqual(result, expect)
                     return lambda result: None
+
                 @jit
                 def check_eq(x, y):
                     bar(operator(x, y))
+
                 check_eq(x, y)
 
         check(operator.ne)

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1180,23 +1180,52 @@ class TestOperatorMixedTypes(TestCase):
 
     def test_eq_ne_returns_literal_on_none(self):
 
-        def bar():
+        def bar(d, e):
             ...
 
-        @overload(bar)
-        def ol_bar(d):
-            # Reject non-literal types to force retry with literal
-            if not isinstance(d, types.Literal):
-                return None
+        @overload(bar, prefer_literal=True)
+        def ol_bar(d, e):
             # Assert that the result received is a literal
             self.assertEqual(hasattr(d, 'literal_value'), True)
-            return lambda d: d
+            # Assert that the result received has the correct literal_value
+            self.assertEqual(d.literal_value, e.literal_value)
+            return lambda d, e: d
 
         @njit
-        def eq():
-            return bar(1)
+        def eq_none_none():
+            return bar(None == None, True)
 
-        self.assertTrue(eq())
+        self.assertTrue(eq_none_none())
+
+        @njit
+        def ne_none_none():
+            return bar(None != None, False)
+
+        self.assertFalse(ne_none_none())
+
+        @njit
+        def eq_none_int():
+            return bar(None == 1, False)
+
+        self.assertFalse(eq_none_int())
+
+        @njit
+        def ne_none_int():
+            return bar(None != 1, True)
+
+        self.assertTrue(ne_none_int())
+
+        @njit
+        def eq_int_none():
+            return bar(1 == None, False)
+
+        self.assertFalse(eq_int_none())
+
+        @njit
+        def ne_int_none():
+            return bar(1 != None, True)
+
+        self.assertTrue(ne_int_none())
 
 
     def test_10414(self):

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -13,7 +13,8 @@ from numba import jit, typeof, njit, typed
 from numba.core import errors, types, config
 from numba.tests.support import (TestCase, tag, ignore_internal_warnings,
                                  MemoryLeakMixin)
-from numba.core.extending import overload_method, box, register_jitable
+from numba.core.extending import (overload_method, box, register_jitable,
+                                  overload)
 
 
 forceobj_flags = {'forceobj': True}
@@ -1177,27 +1178,40 @@ class TestOperatorMixedTypes(TestCase):
             for x, y in itertools.product(things, things):
                 self.assertPreciseEqual(func.py_func(x, y), func(x, y))
 
+    def test_eq_ne_returns_literal_on_none(self):
+
+        def bar():
+            ...
+
+        @overload(bar)
+        def ol_bar(d):
+            # Reject non-literal types to force retry with literal
+            if not isinstance(d, types.Literal):
+                return None
+            # Assert that the result received is a literal
+            self.assertEqual(hasattr(d, 'literal_value'), True)
+            return lambda d: d
+
+        @njit
+        def eq():
+            return bar(1)
+
+        self.assertTrue(eq())
+
+
     def test_10414(self):
         # test for https://github.com/numba/numba/issues/10414
         @njit
-        def func_eq():
+        def func(x):
             a = None
-            for _ in range(0):
-                a = 'b'
-            # a is now an OptionalType(unicode_type) -- we test that equals
-            # works correctly.
-            return a == None
-        self.assertEqual(func_eq(), func_eq.py_func())
+            if x:
+                a = 0
+            # a is an OptionalType(unicode_type) if x is False -- we test that
+            # equals and not-equals works correctly.
+            return a == None, a != None
+        self.assertEqual(func(True), func.py_func(True))
+        self.assertEqual(func(False), func.py_func(False))
 
-        @njit
-        def func_ne():
-            a = None
-            for _ in range(0):
-                a = 'b'
-            # a is now an OptionalType(unicode_type) -- we test that not equals
-            # works correctly.
-            return a != None
-        self.assertEqual(func_ne(), func_ne.py_func())
 
     def test_cmp(self):
         for opstr in ('gt', 'lt', 'ge', 'le', 'eq', 'ne'):

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1180,7 +1180,7 @@ class TestOperatorMixedTypes(TestCase):
 
     def test_eq_ne_returns_literal_on_none(self):
 
-        counter = []
+        counter = [0]
 
         def bar(d, e, c):
             ...
@@ -1191,7 +1191,7 @@ class TestOperatorMixedTypes(TestCase):
             self.assertEqual(hasattr(d, 'literal_value'), True)
             # Assert that the result received has the correct literal_value
             self.assertEqual(d.literal_value, e.literal_value)
-            counter.append(True)
+            counter[0] += 1
             return lambda d, e, c: d
 
         @njit
@@ -1231,7 +1231,7 @@ class TestOperatorMixedTypes(TestCase):
         self.assertTrue(ne_int_none())
 
         # Assert that the overload was called a total of 6 times.
-        self.assertEqual(6, len(counter))
+        self.assertEqual(6, counter[0])
 
 
     def test_10414(self):

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1207,17 +1207,16 @@ class TestOperatorMixedTypes(TestCase):
 
     def test_10414(self):
         # test for https://github.com/numba/numba/issues/10414
+
         @njit
         def func(x):
             a = None
             if x:
                 a = 0
-            # a is an OptionalType(int) if x is False -- we test that
-            # equals and not-equals works correctly.
             return a == None, a != None
+
         self.assertEqual(func(True), func.py_func(True))
         self.assertEqual(func(False), func.py_func(False))
-
 
     def test_cmp(self):
         for opstr in ('gt', 'lt', 'ge', 'le', 'eq', 'ne'):

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1241,7 +1241,7 @@ class TestOperatorMixedTypes(TestCase):
             a = None
             if x:
                 a = 0
-            # a is an OptionalType(unicode_type) if x is False -- we test that
+            # a is an OptionalType(int) if x is False -- we test that
             # equals and not-equals works correctly.
             return a == None, a != None
         self.assertEqual(func(True), func.py_func(True))

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1188,7 +1188,9 @@ class TestOperatorMixedTypes(TestCase):
         @overload(bar, prefer_literal=True)
         def ol_bar(d, e, c):
             # Assert that the result received is a literal
-            self.assertEqual(hasattr(d, 'literal_value'), True)
+            self.assertIsInstance(d, types.Literal)
+            # Assert that the result received has the 'literal_value' attribute
+            self.assertTrue(hasattr(d, 'literal_value'))
             # Assert that the result received has the correct literal_value
             self.assertEqual(d.literal_value, e.literal_value)
             counter[0] += 1

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1192,8 +1192,6 @@ class TestOperatorMixedTypes(TestCase):
                     expect = types.literal(operator(x,y))
                     # Assert the result is of type types.Literal
                     self.assertIsInstance(result, types.Literal)
-                    # Assert that result has 'literal_value' attribute
-                    self.assertTrue(hasattr(result, 'literal_value'))
                     # Assert that result has the correct literal_value
                     self.assertEqual(result, expect)
                     return lambda result: None

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1180,52 +1180,58 @@ class TestOperatorMixedTypes(TestCase):
 
     def test_eq_ne_returns_literal_on_none(self):
 
-        def bar(d, e):
+        counter = []
+
+        def bar(d, e, c):
             ...
 
         @overload(bar, prefer_literal=True)
-        def ol_bar(d, e):
+        def ol_bar(d, e, c):
             # Assert that the result received is a literal
             self.assertEqual(hasattr(d, 'literal_value'), True)
             # Assert that the result received has the correct literal_value
             self.assertEqual(d.literal_value, e.literal_value)
-            return lambda d, e: d
+            counter.append(True)
+            return lambda d, e, c: d
 
         @njit
         def eq_none_none():
-            return bar(None == None, True)
+            return bar(None == None, True, 0)
 
         self.assertTrue(eq_none_none())
 
         @njit
         def ne_none_none():
-            return bar(None != None, False)
+            return bar(None != None, False, 1)
 
         self.assertFalse(ne_none_none())
 
         @njit
         def eq_none_int():
-            return bar(None == 1, False)
+            return bar(None == 1, False, 2)
 
         self.assertFalse(eq_none_int())
 
         @njit
         def ne_none_int():
-            return bar(None != 1, True)
+            return bar(None != 1, True, 3)
 
         self.assertTrue(ne_none_int())
 
         @njit
         def eq_int_none():
-            return bar(1 == None, False)
+            return bar(1 == None, False, 4)
 
         self.assertFalse(eq_int_none())
 
         @njit
         def ne_int_none():
-            return bar(1 != None, True)
+            return bar(1 != None, True, 5)
 
         self.assertTrue(ne_int_none())
+
+        # Assert that the overload was called a total of 6 times.
+        self.assertEqual(6, len(counter))
 
 
     def test_10414(self):

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1190,8 +1190,6 @@ class TestOperatorMixedTypes(TestCase):
                 def ol_bar(result):
                     # compile expected result
                     expect = types.literal(operator(x,y))
-                    # Assert the result is of type types.Literal
-                    self.assertIsInstance(result, types.Literal)
                     # Assert that result has the correct literal_value
                     self.assertEqual(result, expect)
                     return lambda result: None

--- a/numba/tests/test_builtins.py
+++ b/numba/tests/test_builtins.py
@@ -1180,61 +1180,28 @@ class TestOperatorMixedTypes(TestCase):
 
     def test_eq_ne_returns_literal_on_none(self):
 
-        counter = [0]
+        def check(operator):
+            for (x, y) in ((None, None), (None, 1), (1, None)):
+                def bar(result):
+                    pass
+                @overload(bar, prefer_literal=True)
+                def ol_bar(result):
+                    # compile expected result
+                    expect = types.literal(operator(x,y))
+                    # Assert the result is of type types.Literal
+                    self.assertIsInstance(result, types.Literal)
+                    # Assert that result has 'literal_value' attribute
+                    self.assertTrue(hasattr(result, 'literal_value'))
+                    # Assert that result has the correct literal_value
+                    self.assertEqual(result, expect)
+                    return lambda result: None
+                @jit
+                def check_eq(x, y):
+                    bar(operator(x, y))
+                check_eq(x, y)
 
-        def bar(d, e, c):
-            ...
-
-        @overload(bar, prefer_literal=True)
-        def ol_bar(d, e, c):
-            # Assert that the result received is a literal
-            self.assertIsInstance(d, types.Literal)
-            # Assert that the result received has the 'literal_value' attribute
-            self.assertTrue(hasattr(d, 'literal_value'))
-            # Assert that the result received has the correct literal_value
-            self.assertEqual(d.literal_value, e.literal_value)
-            counter[0] += 1
-            return lambda d, e, c: d
-
-        @njit
-        def eq_none_none():
-            return bar(None == None, True, 0)
-
-        self.assertTrue(eq_none_none())
-
-        @njit
-        def ne_none_none():
-            return bar(None != None, False, 1)
-
-        self.assertFalse(ne_none_none())
-
-        @njit
-        def eq_none_int():
-            return bar(None == 1, False, 2)
-
-        self.assertFalse(eq_none_int())
-
-        @njit
-        def ne_none_int():
-            return bar(None != 1, True, 3)
-
-        self.assertTrue(ne_none_int())
-
-        @njit
-        def eq_int_none():
-            return bar(1 == None, False, 4)
-
-        self.assertFalse(eq_int_none())
-
-        @njit
-        def ne_int_none():
-            return bar(1 != None, True, 5)
-
-        self.assertTrue(ne_int_none())
-
-        # Assert that the overload was called a total of 6 times.
-        self.assertEqual(6, counter[0])
-
+        check(operator.ne)
+        check(operator.eq)
 
     def test_10414(self):
         # test for https://github.com/numba/numba/issues/10414

